ecosystem/sep-0039.md

```diff
@@ -0,0 +1,182 @@
+
+Note also that this SEP does not address concerns around *immutability*. While there is an optional checksum component (addressed later), the account itself is free to modify the data at-will if it is not locked.
+
+Our exclusive focus in this SEP is Step 2: **storing the asset within the account.**
+
+### Data Format
+We begin by describing the raw data format, then dive into each component.
+
+    00|version|metadata length|media-type;param1=value;param2=value;etc<binary data>
+
+Here, `|` is not a character, but rather a semantic divider used to separate the components. The following sections break down each component in detail.
+
+#### Index
+Since an account can only have 1000 entries, we only need 10 bits for an index field. We need indices to be ASCII-sortable in Horizon, so the index is expressed as a "base 36" value: 0-9 then a-z.
+
+A valid SEP-39 entry always begins with `00`, indicating that this is the 0th row of data. In fact, every row's key _must_ begin with a two-character, [Base36] index.
+
+---------
+
+**Author's note**: This portion is definitely up for debate. Base36 (well, more accurately, a radix of 36) allows us to shave a byte off of each row (saving up to 1kB total), but we could stick to a 000-999 model (stringified integer) for simplicity. Note that supporting it is trivial in any language, however, since we know it's strictly an integer. For example,
+
+```python
+import string 
+ALPHABET = string.digits + string.ascii_lowercase
+
+def int_to_b36(i: int) -> str:
+    d, r = divmod(i, len(ALPHABET))
+    return ALPHABET[d] + ALPHABET[r]
+
+def b36_to_int(s: str) -> int:
+    return ALPHABET.index(s[0]) * len(ALPHABET) + ALPHABET.index(s[1])
+```
+
+Another consideration is perhaps trying to have **two** indices within this value. Since we need 10 bits, we can use the other 6 for identifying this asset within the account itself. There's still the question of printables, but  though, so this needs more thought.
+
+---------
+
+#### Version
+The first entry's index is followed by the version string. For the current version of this SEP, this is just `1`.
+
+#### Metadata
+The **metadata** is a [Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml) alongside any associated parameters. It is prefixed by a **length**, which is just a stringified integer (that is, a 32-byte metadata will have "32" as its length). When parsing, you should consume characters until hitting a non-digit, since a valid Media Type will never start with a digit. Note that row indices are not included in the metadata length.
+
+**Author's note**: If we go with a radix of 36, we could use that here, too.
+
+There are no *required* parameters.
+
+Standard *optional* parameters include:
```

Comment:
Ah I think I see you're keeping the <binary data> limited to _just_ image data vs title, description and image data broken up by using the `len` like we had previously. I don't like this. At least not if the idea is that this is a raw string. Descriptions can be quite long so this at _least_ needs to be encoded into the basE91 at the very least. I still think I like the binary concatenation with the `len` option to allow other buffers to be encoded into a super buffer. This would also allow support for multiple non-utf8 assets to be encoded into a single sep39 account.

Reply:
Going to respond to [this](https://github.com/stellar/stellar-protocol/pull/1090#discussion_r760707194) here as they're related, I think.

> Descriptions can be quite long so this at least needs to be encoded into the basE91 at the very least.

As I was defining the metadata format, I had trouble envisioning a world in which you have non-printable metadata. In that case, why wouldn't it just be part of the header? That is, what's wrong with `image/png;n=A title;d=This is the description.`? Why do they need to be encoded? A long description is still just printable ASCII; encoding it would just be unnecessary blowup. I guess you can't include Media-Type identifiers, like `;` or `=` in your description... but that's a bit of a weak argument.

Maybe it's a little wonky if you start getting cute with emojis or something, but the beauty of the metadata format is that at that point, you're free to do what you want, right? Anyone can define the multi-buffer interface on top of this themselves. For example, have the metadata be `image/png;n=Title;d=12` and just take the first 12 bytes of the returned binary and treat them like your description. _You_ have a specific use case, but does it generalize? Do _we_ need to impose specifics?

To be clear, I do see value in standardizing the multi-buffer approach with lengths in the meta, but I'm still not convinced that it's necessary when someone could define it "on top" of the existing format. Heck, even having a title and description be "official" optional parameters seems unnecessary to me. If you want them, include them.

Reply:
Ahhhh okay, sorry, I went back and read our earlier back-and-forth and I see the disconnect now. To me, there's was single media type and metadata (title, desc) went with it (the fact that it was `text/plain` went :arrow_right: over my head). Hence my suggestion to just use strings, define the layer yourself, etc.

You're saying: **let there be many media types and let the header define them.**

New suggestion, then (if we're sure we want to do this):

`media-type/subtype;s=<size>,media-type-2/subtype;s=<size>`

Size is optional if there's one media type, required otherwise. 

Drop name, description, etc. as optional parameters. Let that be part of the buffer itself because there's no reason for us to impose a particular convention about what a field is "called" (leave that up to the user In other words, if you have a title field, let it just say `text/plain;s=19` in the header, then it as `Title=Hello, world!` (or `title:Hello, world!` or just `Hello, world!` because you implicitly know it's the title) within the binary data itself. 

Point is, we don't need to enforce that in the SEP. Just require a size for multi-part.

Reply:
I think I address this in 383b47c.

Reply:
> Let that be part of the buffer itself

Field titles absolutely should be part of the metadata. If they're baked into the buffer itself I now have to decode the buffer, split off an `=` or `:` then re-buffer to deliver to the client. The data should be only data so I can just deliver that buffer in a REST response. I agree we don't need to enforce any naming conventions. Just maybe have `n` for `name` be a suggested optional value. So we'd have `l` for length and `n` for name and `c` for checksum. If you opt to store the data in the metadata itself you use the `d` key.

Reply:
`l` and `d` then would be mutually exclusive. Either, or or neither in the case of one or the last value in the metadata set of `meta-type`s

Reply:
My rationale was that "what if I don't care about a name, and I already know that the "text/plain" part of the buffer is the name?" But if it's optional, that concern is alleviated anyway.

Regarding `d`, I say we disallow that entirely. Let's require it to be in the buffer; it keeps parsing simple.

Reply:
A strong reason to keep `d` is that for text/plain fields keeping it as plaintext vs converting to basE91 will keep things more efficient. And again in the case of a long description it could cross quite a few keys and thus forcing that to go through the basE91 encoding will lose some bytes.

Reply:
On second thought any `=` `,` or `;` in a metadata `d` fields would decimate the ability to appropriately parse the metadata 🤔  yeah we'd better keep all data, plaintext or otherwise outside the metadata.

Reply:
Good rationale, but there's unfortunately a huge blocker: Media Types technically can't have whitespace :sweat_smile:. I didn't realize this earlier.

Reply:
It's settled then. All data will live in the main binary data buffer with splitting denoted by the `l` key. `l` or `s`? I've seen both.

Reply:
Let's actually use `s` because it's easier to distinguish visually than `l` vs `1`.